--- acpica-unix-20120420/source/compiler/Makefile	2012-04-20 22:02:33.000000000 +0200
+++ acpica-unix-20120420.patched/source/compiler/Makefile	2012-10-21 00:30:38.634470788 +0200
@@ -257,19 +257,11 @@
 # gcc 4+ flags
 #
 CWARNINGFLAGS += \
-    -Waddress \
     -Waggregate-return \
     -Wchar-subscripts \
-    -Wempty-body \
-    -Wlogical-op \
     -Wmissing-declarations \
-    -Wmissing-field-initializers \
-    -Wmissing-parameter-type \
     -Wnested-externs \
-    -Wold-style-declaration \
-    -Wold-style-definition \
-    -Wredundant-decls \
-    -Wtype-limits
+    -Wold-style-definition
 
 #
 # Bison/Flex configuration
