diff --git a/Makefile.in b/Makefile.in
index 9b3a5d75..b8347953 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -370,7 +370,7 @@ YACC = @YACC@
 FLEX = @FLEX@
 LEX = @LEX@
 M4 = @M4@
-MAKEINFO = @MAKEINFO@
+MAKEINFO = true
 EXPECT = @EXPECT@
 RUNTEST = @RUNTEST@
 
diff --git a/Makefile.tpl b/Makefile.tpl
index 84fee3dd..79a3ea7a 100644
--- a/Makefile.tpl
+++ b/Makefile.tpl
@@ -373,7 +373,7 @@ YACC = @YACC@
 FLEX = @FLEX@
 LEX = @LEX@
 M4 = @M4@
-MAKEINFO = @MAKEINFO@
+MAKEINFO = true
 EXPECT = @EXPECT@
 RUNTEST = @RUNTEST@
 
diff --git a/bfd/Makefile.in b/bfd/Makefile.in
index 12807d99..c42976cb 100644
--- a/bfd/Makefile.in
+++ b/bfd/Makefile.in
@@ -343,7 +343,7 @@ LIPO = @LIPO@
 LN_S = @LN_S@
 LTLIBOBJS = @LTLIBOBJS@
 MAINT = @MAINT@
-MAKEINFO = @MAKEINFO@
+MAKEINFO = true
 MKDIR_P = @MKDIR_P@
 MKINSTALLDIRS = @MKINSTALLDIRS@
 MSGFMT = @MSGFMT@
diff --git a/bfd/doc/Makefile.in b/bfd/doc/Makefile.in
index 0724b778..ff6cf159 100644
--- a/bfd/doc/Makefile.in
+++ b/bfd/doc/Makefile.in
@@ -285,7 +285,7 @@ LIPO = @LIPO@
 LN_S = @LN_S@
 LTLIBOBJS = @LTLIBOBJS@
 MAINT = @MAINT@
-MAKEINFO = @MAKEINFO@
+MAKEINFO = true
 MKDIR_P = @MKDIR_P@
 MKINSTALLDIRS = @MKINSTALLDIRS@
 MSGFMT = @MSGFMT@
diff --git a/binutils/Makefile.in b/binutils/Makefile.in
index bed8fef6..4066d24f 100644
--- a/binutils/Makefile.in
+++ b/binutils/Makefile.in
@@ -469,7 +469,7 @@ LN_S = @LN_S@
 LTLIBICONV = @LTLIBICONV@
 LTLIBOBJS = @LTLIBOBJS@
 MAINT = @MAINT@
-MAKEINFO = @MAKEINFO@
+MAKEINFO = true
 MKDIR_P = @MKDIR_P@
 MKINSTALLDIRS = @MKINSTALLDIRS@
 MSGFMT = @MSGFMT@
diff --git a/binutils/doc/Makefile.in b/binutils/doc/Makefile.in
index 024a642e..c66d21de 100644
--- a/binutils/doc/Makefile.in
+++ b/binutils/doc/Makefile.in
@@ -302,7 +302,7 @@ LN_S = @LN_S@
 LTLIBICONV = @LTLIBICONV@
 LTLIBOBJS = @LTLIBOBJS@
 MAINT = @MAINT@
-MAKEINFO = @MAKEINFO@
+MAKEINFO = true
 MKDIR_P = @MKDIR_P@
 MKINSTALLDIRS = @MKINSTALLDIRS@
 MSGFMT = @MSGFMT@
diff --git a/gas/Makefile.in b/gas/Makefile.in
index f3d66b8e..5a53a9cb 100644
--- a/gas/Makefile.in
+++ b/gas/Makefile.in
@@ -319,7 +319,7 @@ LIPO = @LIPO@
 LN_S = @LN_S@
 LTLIBOBJS = @LTLIBOBJS@
 MAINT = @MAINT@
-MAKEINFO = @MAKEINFO@
+MAKEINFO = true
 MKDIR_P = @MKDIR_P@
 MKINSTALLDIRS = @MKINSTALLDIRS@
 MSGFMT = @MSGFMT@
diff --git a/gas/doc/Makefile.in b/gas/doc/Makefile.in
index 89d2ca06..d98a8883 100644
--- a/gas/doc/Makefile.in
+++ b/gas/doc/Makefile.in
@@ -281,7 +281,7 @@ LIPO = @LIPO@
 LN_S = @LN_S@
 LTLIBOBJS = @LTLIBOBJS@
 MAINT = @MAINT@
-MAKEINFO = @MAKEINFO@
+MAKEINFO = true
 MKDIR_P = @MKDIR_P@
 MKINSTALLDIRS = @MKINSTALLDIRS@
 MSGFMT = @MSGFMT@
diff --git a/gold/Makefile.in b/gold/Makefile.in
index 9fd21214..52e5f365 100644
--- a/gold/Makefile.in
+++ b/gold/Makefile.in
@@ -558,7 +558,7 @@ LIBS = @LIBS@
 LN_S = @LN_S@
 LTLIBOBJS = @LTLIBOBJS@
 MAINT = @MAINT@
-MAKEINFO = @MAKEINFO@
+MAKEINFO = true
 MERGE_CONSTANTS_FLAG = @MERGE_CONSTANTS_FLAG@
 MKDIR_P = @MKDIR_P@
 MKINSTALLDIRS = @MKINSTALLDIRS@
diff --git a/gold/testsuite/Makefile.in b/gold/testsuite/Makefile.in
index 7b4b7832..090a555f 100644
--- a/gold/testsuite/Makefile.in
+++ b/gold/testsuite/Makefile.in
@@ -2644,7 +2644,7 @@ LIBS = @LIBS@
 LN_S = @LN_S@
 LTLIBOBJS = @LTLIBOBJS@
 MAINT = @MAINT@
-MAKEINFO = @MAKEINFO@
+MAKEINFO = true
 MERGE_CONSTANTS_FLAG = @MERGE_CONSTANTS_FLAG@
 MKDIR_P = @MKDIR_P@
 MKINSTALLDIRS = @MKINSTALLDIRS@
diff --git a/gprof/Makefile.in b/gprof/Makefile.in
index 0f766a36..b8c07e2b 100644
--- a/gprof/Makefile.in
+++ b/gprof/Makefile.in
@@ -351,7 +351,7 @@ LIPO = @LIPO@
 LN_S = @LN_S@
 LTLIBOBJS = @LTLIBOBJS@
 MAINT = @MAINT@
-MAKEINFO = @MAKEINFO@
+MAKEINFO = true
 MKDIR_P = @MKDIR_P@
 MKINSTALLDIRS = @MKINSTALLDIRS@
 MSGFMT = @MSGFMT@
diff --git a/ld/Makefile.in b/ld/Makefile.in
index ef2e99e0..9775d3d2 100644
--- a/ld/Makefile.in
+++ b/ld/Makefile.in
@@ -428,7 +428,7 @@ LIPO = @LIPO@
 LN_S = @LN_S@
 LTLIBOBJS = @LTLIBOBJS@
 MAINT = @MAINT@
-MAKEINFO = @MAKEINFO@
+MAKEINFO = true
 MKDIR_P = @MKDIR_P@
 MKINSTALLDIRS = @MKINSTALLDIRS@
 MSGFMT = @MSGFMT@
diff --git a/libctf/Makefile.in b/libctf/Makefile.in
index 5cfa100f..8b3bf7ab 100644
--- a/libctf/Makefile.in
+++ b/libctf/Makefile.in
@@ -360,7 +360,7 @@ LIPO = @LIPO@
 LN_S = @LN_S@
 LTLIBOBJS = @LTLIBOBJS@
 MAINT = @MAINT@
-MAKEINFO = @MAKEINFO@
+MAKEINFO = true
 MKDIR_P = @MKDIR_P@
 NM = @NM@
 NMEDIT = @NMEDIT@
diff --git a/libiberty/Makefile.in b/libiberty/Makefile.in
index d19e14a2..6b1dd3d7 100644
--- a/libiberty/Makefile.in
+++ b/libiberty/Makefile.in
@@ -56,7 +56,7 @@ CC = @CC@
 CFLAGS = @CFLAGS@
 CPPFLAGS = @CPPFLAGS@
 RANLIB = @RANLIB@ @RANLIB_PLUGIN_OPTION@
-MAKEINFO = @MAKEINFO@
+MAKEINFO = true
 PERL = @PERL@
 
 PICFLAG = @PICFLAG@
diff --git a/opcodes/Makefile.in b/opcodes/Makefile.in
index 42c15f00..82b8dc9a 100644
--- a/opcodes/Makefile.in
+++ b/opcodes/Makefile.in
@@ -323,7 +323,7 @@ LIPO = @LIPO@
 LN_S = @LN_S@
 LTLIBOBJS = @LTLIBOBJS@
 MAINT = @MAINT@
-MAKEINFO = @MAKEINFO@
+MAKEINFO = true
 MKDIR_P = @MKDIR_P@
 MKINSTALLDIRS = @MKINSTALLDIRS@
 MSGFMT = @MSGFMT@
diff --git a/zlib/Makefile.in b/zlib/Makefile.in
index 062dc04e..4abb5cee 100644
--- a/zlib/Makefile.in
+++ b/zlib/Makefile.in
@@ -304,7 +304,7 @@ LIPO = @LIPO@
 LN_S = @LN_S@
 LTLIBOBJS = @LTLIBOBJS@
 MAINT = @MAINT@
-MAKEINFO = @MAKEINFO@
+MAKEINFO = true
 MKDIR_P = @MKDIR_P@
 NM = @NM@
 NMEDIT = @NMEDIT@
