diff -ur binutils-2.25/bfd/Makefile.in binutils-2.25.patched/bfd/Makefile.in
--- binutils-2.25/bfd/Makefile.in	2015-09-15 06:25:42.000000000 -0700
+++ binutils-2.25.patched/bfd/Makefile.in	2015-09-15 05:51:01.000000000 -0700
@@ -339,7 +339,7 @@
 ACLOCAL_AMFLAGS = -I . -I .. -I ../config
 INCDIR = $(srcdir)/../include
 CSEARCH = -I. -I$(srcdir) -I$(INCDIR)
-SUBDIRS = doc po
+SUBDIRS = po
 bfddocdir = doc
 libbfd_la_LDFLAGS = $(am__append_1) -release `cat libtool-soversion` \
 	@SHARED_LDFLAGS@ $(am__empty)
