--- binutils-2.18.50/gas/config/tc-i386.c~	2009-01-20 14:19:04.000000000 +0100
+++ binutils-2.18.50/gas/config/tc-i386.c	2009-01-20 14:20:47.000000000 +0100
@@ -183,6 +183,7 @@
 	 && !defined (TE_GNU)				\
 	 && !defined (TE_LINUX)				\
  	 && !defined (TE_NETWARE)			\
+	 && 0						\
 	 && !defined (TE_FreeBSD)			\
 	 && !defined (TE_NetBSD)))
 /* This array holds the chars that always start a comment.  If the
