diff -ur binutils-2.32/bfd/Makefile.in binutils-2.32.patched/bfd/Makefile.in
--- binutils-2.32/bfd/Makefile.in
+++ binutils-2.32.patched/bfd/Makefile.in
@@ -459,7 +459,7 @@
 ACLOCAL_AMFLAGS = -I . -I .. -I ../config
 INCDIR = $(srcdir)/../include
 CSEARCH = -I. -I$(srcdir) -I$(INCDIR)
-SUBDIRS = doc po
+SUBDIRS = po
 bfddocdir = doc
 libbfd_la_LDFLAGS = $(am__append_1) -release `cat libtool-soversion` \
 	@SHARED_LDFLAGS@ $(am__empty)
