diff -ur binutils-2.23.2/bfd/Makefile.in binutils-2.23.2.patched/bfd/Makefile.in
--- binutils-2.23.2/bfd/Makefile.in	2012-10-22 14:33:51.000000000 +0200
+++ binutils-2.23.2.patched/bfd/Makefile.in	2012-10-24 15:33:04.442080163 +0200
@@ -323,7 +323,7 @@
 RELEASE = y
 INCDIR = $(srcdir)/../include
 CSEARCH = -I. -I$(srcdir) -I$(INCDIR)
-SUBDIRS = doc po
+SUBDIRS = po
 bfddocdir = doc
 libbfd_la_LDFLAGS = $(am__append_1) -release `cat libtool-soversion` \
 	@SHARED_LDFLAGS@ $(am__empty)
