--- gcc-6.1.0/gcc/ada/gcc-interface/Make-lang.in.bak	2015-08-24 16:23:25.004493665 +0200
+++ gcc-6.1.0/gcc/ada/gcc-interface/Make-lang.in	2015-08-24 17:53:52.496636113 +0200
@@ -45,7 +45,7 @@
 
 
 # Extra flags to pass to recursive makes.
-COMMON_ADAFLAGS= -gnatpg
+COMMON_ADAFLAGS= -gnatpg -gnatwn
 ifeq ($(TREECHECKING),)
 CHECKING_ADAFLAGS=
 else
diff -Nurp gcc-11.2.0/gcc/ada/gcc-interface/Make-lang.in gcc-11.2.0.new/gcc/ada/gcc-interface/Make-lang.in
--- gcc-11.2.0/gcc/ada/gcc-interface/Make-lang.in	2022-06-03 00:31:57.993273717 +0200
+++ gcc-11.2.0.new/gcc/ada/gcc-interface/Make-lang.in	2022-06-03 00:30:50.214166847 +0200
@@ -334,6 +334,7 @@ GNAT_ADA_OBJS =	\
  ada/hostparm.o	\
  ada/impunit.o	\
  ada/inline.o	\
+ ada/libgnat/i-c.o	\
  ada/libgnat/interfac.o	\
  ada/itypes.o	\
  ada/krunch.o	\
@@ -364,7 +365,10 @@ GNAT_ADA_OBJS =	\
  ada/rtsfind.o	\
  ada/libgnat/s-addope.o	\
  ada/libgnat/s-addima.o \
+ ada/libgnat/s-aotase.o	\
  ada/libgnat/s-assert.o	\
+ ada/libgnat/s-atoope.o	\
+ ada/libgnat/s-atopri.o	\
  ada/libgnat/s-bitops.o	\
  ada/libgnat/s-carun8.o	\
  ada/libgnat/s-casuti.o	\
@@ -548,6 +552,7 @@ GNATBIND_OBJS = \
  ada/hostparm.o   \
  ada/init.o       \
  ada/initialize.o \
+ ada/libgnat/i-c.o        \
  ada/libgnat/interfac.o   \
  ada/krunch.o     \
  ada/lib.o        \
