diff -ur gcc-4.7.3/config.sub gcc-4.7.3-patched/config.sub
--- gcc-4.7.3/config.sub	2012-08-06 16:34:27.000000000 +0200
+++ gcc-4.7.3-patched/config.sub	2013-04-19 00:25:16.401391381 +0200
@@ -250,7 +250,7 @@
 	| alpha | alphaev[4-8] | alphaev56 | alphaev6[78] | alphapca5[67] \
 	| alpha64 | alpha64ev[4-8] | alpha64ev56 | alpha64ev6[78] | alpha64pca5[67] \
 	| am33_2.0 \
-	| arc | arm | arm[bl]e | arme[lb] | armv[2345] | armv[345][lb] | avr | avr32 \
+	| arc | arm | arm[bl]e | arme[lb] | armv* | armv[345][lb] | avr | avr32 \
         | be32 | be64 \
 	| bfin \
 	| c4x | clipper \
