--- t/gcc-4.4.2/gcc/crtstuff.c	Fr. Apr 10 01:23:07 2009
+++ gcc-4.4.2/gcc/crtstuff.c	Di. Jul 28 16:43:28 2009
@@ -204,6 +204,7 @@
   = { (func_ptr) (-1) };
 #endif /* __DTOR_LIST__ alternatives */

+#if 0
 #ifdef USE_EH_FRAME_REGISTRY
 /* Stick a label at the beginning of the frame unwind info so we can register
    and deregister it with the exception handling library code.  */
@@ -219,6 +220,7 @@
   __attribute__ ((unused, section(JCR_SECTION_NAME), aligned(sizeof(void*))))
   = { };
 #endif /* JCR_SECTION_NAME */
+#endif

 #if defined(INIT_SECTION_ASM_OP) || defined(INIT_ARRAY_SECTION_ASM_OP)

@@ -309,6 +311,7 @@
   }
 #endif /* !defined(FINI_ARRAY_SECTION_ASM_OP) */

+#if 0
 #ifdef USE_EH_FRAME_REGISTRY
 #ifdef CRT_GET_RFIB_DATA
   /* If we used the new __register_frame_info_bases interface,
@@ -320,6 +323,7 @@
     __deregister_frame_info (__EH_FRAME_BEGIN__);
 #endif
 #endif
+#endif

   completed = 1;
 }
@@ -333,6 +337,7 @@
   = { __do_global_dtors_aux };
 #endif /* !defined(FINI_SECTION_ASM_OP) */

+#if 0
 #if defined(USE_EH_FRAME_REGISTRY) || defined(JCR_SECTION_NAME)
 /* Stick a call to __register_frame_info into the .init section.  For some
    reason calls with no arguments work more reliably in .init, so stick the
@@ -364,6 +369,7 @@
     }
 #endif /* JCR_SECTION_NAME */
 }
+#endif

 #ifdef INIT_SECTION_ASM_OP
 CRT_CALL_STATIC_FUNCTION (INIT_SECTION_ASM_OP, frame_dummy)
