diff -urN gdb-8.2.1.orig/gdb/python/python.c gdb-8.2.1/gdb/python/python.c
--- gdb-8.2.1.org/gdb/python/python.c	2019-01-04 12:48:32.440687758 +0100
+++ gdb-8.2.1/gdb/python/python.c	2019-01-04 12:54:09.917677688 +0100
@@ -1727,6 +1727,15 @@
 #endif
 #endif
 
+  char readlinkbuffer[BUFSIZ];
+  int readlinks = readlink("/proc/self/exe", readlinkbuffer, BUFSIZ - 1);
+  readlinkbuffer[readlinks] = 0;
+  char *executeablepath = dirname(readlinkbuffer);
+  char *pythonhome = malloc(strlen(executeablepath) + strlen("/../") + 2);
+  strcpy(pythonhome, executeablepath);
+  strcat(pythonhome, "/../");
+  setenv("PYTHONHOME", pythonhome, 1);
+
   Py_Initialize ();
   PyEval_InitThreads ();
 
