diff -urN gdb-3.orig/gdb/configure.tgt gdb-8.3/gdb/configure.tgt
--- gdb-8.3.orig/gdb/configure.tgt	2019-02-27 05:51:48.000000000 +0100
+++ gdb-8.3/gdb/configure.tgt	2019-05-12 08:07:45.624984535 +0200
@@ -772,6 +772,10 @@
 x86_64-*-rtems*)
 	gdb_target_obs="${amd64_tobjs} ${i386_tobjs} i386-bsd-tdep.o"
 	;;
+x86_64-*-*)
+	# Target: amd6
+	gdb_target_obs="amd64-tdep.o i386-tdep.o i387-tdep.o"
+	;;
 xtensa*-*-*linux*)
 	# Target: GNU/Linux Xtensa
 	gdb_target_obs="xtensa-linux-tdep.o symfile-mem.o linux-tdep.o"
