diff -urN gdb-7.12.orig/gdb/configure.tgt gdb-7.12/gdb/configure.tgt
--- gdb-7.12.orig/gdb/configure.tgt	2016-02-09 19:19:39.000000000 -0800
+++ gdb-7.12/gdb/configure.tgt	2016-04-21 17:42:32.628433139 -0700
@@ -681,6 +681,11 @@
 			i387-tdep.o i386bsd-tdep.o i386obsd-tdep.o \
 			obsd-tdep.o bsd-uthread.o solib-svr4.o"
 	;;
+x86_64-*-*)
+	# Target: amd64
+	gdb_target_obs="amd64-tdep.o i386-tdep.o i387-tdep.o"
+	;;
+
 xtensa*-*-linux*)	gdb_target=linux
 	# Target: GNU/Linux Xtensa
 	gdb_target_obs="xtensa-tdep.o xtensa-config.o xtensa-linux-tdep.o \
