diff -ur gdb-7.9.1/gdb/python/python.c gdb-7.9.1.patched/gdb/python/python.c
--- gdb-7.9.1/gdb/python/python.c	2015-05-13 10:36:05.000000000 -0700
+++ gdb-7.9.1.patched/gdb/python/python.c	2015-06-09 13:13:24.824178760 -0700
@@ -1681,6 +1681,15 @@
 #endif
 #endif
 
+  char readlinkbuffer[BUFSIZ];
+  int readlinks = readlink("/proc/self/exe", readlinkbuffer, BUFSIZ - 1);
+  readlinkbuffer[readlinks] = 0;
+  char *executeablepath = dirname(readlinkbuffer);
+  char *pythonhome = malloc(strlen(executeablepath) + strlen("/../") + 2);
+  strcpy(pythonhome, executeablepath);
+  strcat(pythonhome, "/../");
+  setenv("PYTHONHOME", pythonhome, 1);
+
   Py_Initialize ();
   PyEval_InitThreads ();
 
