diff -urN gdb-7.12.orig/gdb/Makefile.in gdb-7.12/gdb/Makefile.in
--- gdb-7.12.orig/gdb/Makefile.in	2016-02-24 01:55:15.000000000 -0800
+++ gdb-7.12/gdb/Makefile.in	2016-04-21 17:44:32.721472633 -0700
@@ -1092,7 +1092,7 @@
 
 TSOBS = inflow.o
 
-SUBDIRS = doc @subdirs@ data-directory $(GNULIB_BUILDDIR)
+SUBDIRS = @subdirs@ data-directory $(GNULIB_BUILDDIR)
 CLEANDIRS = $(SUBDIRS)
 
 # List of subdirectories in the build tree that must exist.
