diff -urN gdb-8.0.orig/gdb/Makefile.in gdb-8.0/gdb/Makefile.in
--- gdb-8.2.1.orig/gdb/Makefile.in	2019-01-04 12:58:10.227572419 +0100
+++ gdb-8.2.1/gdb/Makefile.in	2019-01-04 12:59:00.539134274 +0100
@@ -1568,7 +1568,7 @@
 
 TSOBS = inflow.o
 
-SUBDIRS = doc @subdirs@ data-directory $(GNULIB_BUILDDIR)
+SUBDIRS = @subdirs@ data-directory $(GNULIB_BUILDDIR)
 CLEANDIRS = $(SUBDIRS)
 
 # List of subdirectories in the build tree that must exist.
