--- /tmp/gdb-7.2/gdb/python/python.c	2010-08-09 21:22:59.000000000 +0200
+++ gdb-7.2/gdb/python/python.c	2010-10-28 15:31:35.000000000 +0200
@@ -28,6 +28,7 @@
 #include "value.h"
 #include "language.h"
 #include "exceptions.h"
+#include <libgen.h>

 #include <ctype.h>

@@ -684,6 +685,16 @@
 			     SLASH_STRING, "python", NULL));
 #endif

+
+  char readlinkbuffer[BUFSIZ];
+  int readlinks = readlink("/proc/self/exe", readlinkbuffer, BUFSIZ - 1);
+  readlinkbuffer[readlinks] = 0;
+  char *executeablepath = dirname(readlinkbuffer);
+  char *pythonhome = malloc(strlen(executeablepath) + strlen("/../") + 2);
+  strcpy(pythonhome, executeablepath);
+  strcat(pythonhome, "/../");
+  setenv("PYTHONHOME", pythonhome, 1);
+
   Py_Initialize ();
   PyEval_InitThreads ();

