Index: kconfig/confdata.c
===================================================================
--- kconfig.orig/confdata.c
+++ kconfig/confdata.c
@@ -230,6 +230,13 @@ static const char *conf_get_autoheader_n
 	return name ? name : "include/generated/autoconf.h";
 }
 
+static const char *conf_get_autobase_name(void)
+{
+	char *name = getenv("KCONFIG_SPLITCONFIG");
+
+	return name ? name : "include/config/";
+}
+
 static int conf_set_sym_val(struct symbol *sym, int def, int def_flags, char *p)
 {
 	char *p2;
@@ -1031,7 +1038,7 @@ static int conf_touch_deps(void)
 	struct symbol *sym;
 	int res, i;
 
-	strcpy(depfile_path, "include/config/");
+	strcpy(depfile_path, conf_get_autobase_name());
 	depfile_prefix_len = strlen(depfile_path);
 
 	name = conf_get_autoconfig_name();
