diff -urN linux-2.4.7-official/arch/i386/Linuxbios.in linux-2.4.7-linuxbios/arch/i386/Linuxbios.in
--- linux-2.4.7-official/arch/i386/Linuxbios.in	Thu Jan  1 08:00:00 1970
+++ linux-2.4.7-linuxbios/arch/i386/Linuxbios.in	Fri Jul 27 11:19:47 2001
@@ -0,0 +1,5 @@
+mainmenu_option next_comment
+comment 'Linuxbios Options'
+bool 'Force IDE Controllers on?' CONFIG_LINUXBIOS_FORCE_IDE_CONTROLLER_ON
+bool 'Wait for HDA to spin up?'  CONFIG_LINUXBIOS_WAIT_HDA_SPINUP
+endmenu
diff -urN linux-2.4.7-official/arch/i386/config.in linux-2.4.7-linuxbios/arch/i386/config.in
--- linux-2.4.7-official/arch/i386/config.in	Mon Jul 16 07:15:44 2001
+++ linux-2.4.7-linuxbios/arch/i386/config.in	Fri Jul 27 11:19:47 2001
@@ -391,3 +391,13 @@
 #bool 'Debug kmalloc/kfree' CONFIG_DEBUG_MALLOC
 bool 'Magic SysRq key' CONFIG_MAGIC_SYSRQ
 endmenu
+
+mainmenu_option next_comment
+comment 'LinuxBIOS'
+
+tristate 'LinuxBIOS Support' CONFIG_LINUXBIOS
+if [ "$CONFIG_LINUXBIOS" != "n" ]; then
+   source arch/i386/Linuxbios.in
+fi
+
+endmenu
diff -urN linux-2.4.7-official/drivers/ide/ide-probe.c linux-2.4.7-linuxbios/drivers/ide/ide-probe.c
--- linux-2.4.7-official/drivers/ide/ide-probe.c	Mon Mar 19 01:25:02 2001
+++ linux-2.4.7-linuxbios/drivers/ide/ide-probe.c	Fri Jul 27 11:19:48 2001
@@ -310,6 +310,13 @@
 		if ((drive->media != ide_disk) && (cmd == WIN_IDENTIFY))
 			return 4;
 	}
+#ifdef CONFIG_LINUXBIOS_WAIT_HDA_SPINUP
+	if (! strncmp(drive->name, "hda", 3)) {
+	  printk("jamming drive present for %s\n", drive->name);
+	  drive->present = 1;
+	}
+#endif
+
 #ifdef DEBUG
 	printk("probing for %s: present=%d, media=%d, probetype=%s\n",
 		drive->name, drive->present, drive->media,
