Index: src/northbridge/amd/amdk8/raminit.c
===================================================================
RCS file: /cvsroot/freebios/freebios2/src/northbridge/amd/amdk8/raminit.c,v
retrieving revision 1.26
diff -u -r1.26 raminit.c
--- src/northbridge/amd/amdk8/raminit.c	14 Oct 2004 22:06:29 -0000	1.26
+++ src/northbridge/amd/amdk8/raminit.c	26 Jan 2005 09:50:05 -0000
@@ -1603,7 +1603,7 @@
 	dimm += DCL_x4DIMM_SHIFT;
 	dcl = pci_read_config32(ctrl->f2, DRAM_CONFIG_LOW);
 	dcl &= ~(1 << dimm);
-	if (value == 4) {
+	if (value >= 4) {
 		dcl |= (1 << dimm);
 	}
 	pci_write_config32(ctrl->f2, DRAM_CONFIG_LOW, dcl);
